bpf/lbmap/lbmap.go

```diff
@@ -173,13 +208,94 @@ func LookupRevNat(key RevNatKey) (RevNatValue, error) {
 	return revnat.Convert(), nil
 }
 
+func gcd(x, y uint16) uint16 {
+	for y != 0 {
+		x, y = y, x%y
+	}
+	return x
+}
+
+type rrState struct {
+	index int
+	cw    uint16
+}
+
+func next(state *rrState, n int, max uint16, weights []uint16) uint16 {
+	for {
+		state.index = (state.index + 1) % n
+		if state.index == 0 {
+			state.cw = state.cw - 1
+			if state.cw <= 0 {
+				state.cw = max
+			}
+		}
+		if weights[state.index] >= state.cw {
+			return uint16(state.index)
+		}
+	}
+}
+
+func GenerateWrrSeq(weights []uint16) *ServiceRRSeq {
+	var seq []uint16
+	svcRRSeq := ServiceRRSeq{}
+
+	n := len(weights)
+	g := gcd(weights[0], weights[1])
+
+	if g == 0 {
+		return nil
+	}
+
+	for i := 2; i < n; i++ {
+		g = gcd(g, weights[i])
+	}
+
+	sum := uint16(0)
+	for i := 0; i < n; i++ {
+		weights[i] = weights[i] / g
+		sum += weights[i]
+	}
+
+	state := rrState{
+		index: -1,
+		cw:    uint16(0),
+	}
+	for i := 0; i < int(sum); i++ {
+		seq = append(seq, next(&state, n, sum, weights))
```

Comment:
We should return nil if the len of the sequence does not fit the bpf max so we fall back to hash based. The second loop yeah I can directly write to the array.